### PR TITLE
Acceptors and message structure

### PR DESCRIPTION
**Pull request by dvr0**
I defined a message structure for the Paxos phase variables and added methods to encode and decode an ordered list into a binary message that can be sent over the network. <br>
It can be used to communicate between the actors (proposer, acceptors, learner).

Acceptor function implemented accordingly with a dummy proposer for testing.

### Message structure
Basically an arbitrary number of chunks of 2 bytes each. At the moment only positive integer values in the range 0-65535  are accepted (is it enough?). More details in the code comments.

Example 
- phase 1A: <instance_number><phase_ID><c-rnd>
- phase 2A:  <instance_number><phase_ID><c-rnd><c-val>
where <..> = chunk = 2 bytes.

### Encode and decode
Convert between a "list of chunks" (list of integers) with each chunk corresponding to a separate value that we want to transmit. Order and content depend on the phase.

- encode **phase 1A**:  [1,2,3] -> binary message 
- decode  **phase 1A**:  binary message -> [1,2,3] 

@BritikovKI I changed you learner implementation to reflect the message structure. Let me know if it's fine. Also when you do `if id < len(msg):` is `len(msg)` just used as a number to compare to `id` or does the length actually matter? If it's the first, maybe we could use something smaller that 2**16? :D

### CODE REVIEW

**Comment by dvr0** [comment]
PS: is this the simplest way to do this? probably not lol but I have done too much C lately and it ruined my mind

**Review by jjvegaes — APPROVED** [review]
Great job! Not sure if it is useful the encoder/decoder part, but was needed to put a language in common!

File: our-paxos/paxos.py [hunk 1]
```diff
@@ -28,31 +28,146 @@ def parse_cfg(cfgpath):
...
+        state = paxos_instances[instance]
+        if phase == 1:
+            # received phase 1A msg from proposer
+            if loc[2] > state['rnd']:
+                # update state and paxos_
+                state['rnd'] = loc[2]
+                print("i: {} p: {}, state: {}".format(instance, phase, state))
+                # encode and send phase 1B to proposer
+                msg = paxos_encode([instance, phase, state['rnd'], state['v-rnd'], state['v-val']])
+                s.sendto(msg, config['proposers'])
```
Comment (BritikovKI) [inline our-paxos/paxos.py:130]: We can send message back even if loc[2] == state['rnd']

File: our-paxos/paxos.py [hunk 2]
```diff
@@ -28,31 +28,146 @@ def parse_cfg(cfgpath):
...
+            # loc[2] = c-rnd, loc[3] = c-val
+            if loc[2] >= state['rnd']:
+                #state['rnd'] = loc[2] # not in the slides, but it makes sense (?)
+                state['v-rnd'] = loc[2]
+                state['v-val'] = loc[3]
+                print("i: {} p: {}, state: {}".format(instance, phase, state))
+
+                # encode and send phase 2B to proposer
+                msg = paxos_encode([instance, phase, state['v-rnd'], state['v-val']])
+                s.sendto(msg, config['learners'])
```
Comment (BritikovKI) [inline our-paxos/paxos.py:142]: Same thing as in the previous comment, if I remember correctly we send 2B in any case

File: our-paxos/paxos.py (hunk 1 above)
Comment (BritikovKI) [reply our-paxos/paxos.py:130]: It's not necessary I think, but it will allow to optimise the proposer

**Review by BritikovKI — APPROVED** [review]
Lgtmt

**Pull request merged**